File: bug06_loop_counter_overwritten/main.cpp
#include <iostream>
#include <vector>
#include <cstdint>
#include <cstdlib>

void matrix_hash() {
    uint64_t m[4][4];
    for(int i=0; i<4; i++) for(int j=0; j<4; j++) m[i][j] = 0x123 * (i+1) * (j+1);
    uint64_t total = 0;
    for(uint64_t step = 0; step < 10; step++) {
        uint64_t new_m[4][4] = {0};
        for(int i=0; i<4; i++) {
            for(int j=0; j<4; j++) {
                uint64_t s = 0;
                for(int k=0; k<4; k++) s += m[i][k] * m[k][j];
                new_m[i][j] = s & 0xFFFF;
            }
        }
        for(int i=0; i<4; i++) for(int j=0; j<4; j++) m[i][j] = new_m[i][j];
        
        uint64_t diag = 0;
        for(int i=0; i<4; i++) diag += m[i][i];
        step = diag;
        
        total = (total ^ step) * 0x9E3779B97F4A7C15ULL;
    }

    uint64_t correct = 0;
    uint64_t m2[4][4];
    for(int i=0; i<4; i++) for(int j=0; j<4; j++) m2[i][j] = 0x123 * (i+1) * (j+1);
    for(uint64_t step = 0; step < 10; step++) {
        uint64_t new_m[4][4] = {0};
        for(int i=0; i<4; i++) {
            for(int j=0; j<4; j++) {
                uint64_t s = 0;
                for(int k=0; k<4; k++) s += m2[i][k] * m2[k][j];
                new_m[i][j] = s & 0xFFFF;
            }
        }
        for(int i=0; i<4; i++) for(int j=0; j<4; j++) m2[i][j] = new_m[i][j];
        
        uint64_t diag = 0;
        for(int i=0; i<4; i++) diag += m2[i][i];
        correct = (correct ^ diag) * 0x9E3779B97F4A7C15ULL;
    }

    if (total == correct) {
        std::cerr << "AssertionError: reassigning step gives same result as using diag\n";
        std::exit(1);
    }
    std::cerr << "ValueError: step variable was reassigned: loop counter corrupted, got 0x" << std::hex << total << " expected 0x" << correct << "\n";
    std::exit(1);
}

int main() {
    matrix_hash();
    return 0;
}

File: bug06_loop_counter_overwritten/main.py
import base64, struct

def matrix_hash():
    m = [[0x123 * (i + 1) * (j + 1) for j in range(4)] for i in range(4)]
    total = 0
    for step in range(10):
        new_m = [[0]*4 for _ in range(4)]
        for i in range(4):
            for j in range(4):
                new_m[i][j] = sum(m[i][k] * m[k][j] for k in range(4)) & 0xFFFF
        m = new_m
        step = sum(m[i][i] for i in range(4))
        total = (total ^ step) * 0x9E3779B97F4A7C15
        total &= 0xFFFFFFFFFFFFFFFF

    correct = 0
    m2 = [[0x123 * (i + 1) * (j + 1) for j in range(4)] for i in range(4)]
    for step in range(10):
        new_m = [[0]*4 for _ in range(4)]
        for i in range(4):
            for j in range(4):
                new_m[i][j] = sum(m2[i][k] * m2[k][j] for k in range(4)) & 0xFFFF
        m2 = new_m
        diag = sum(m2[i][i] for i in range(4))
        correct = (correct ^ diag) * 0x9E3779B97F4A7C15
        correct &= 0xFFFFFFFFFFFFFFFF

    if total == correct:
        raise AssertionError("reassigning step gives same result as using diag")
    raise ValueError(f"step variable was reassigned: loop counter corrupted, got {total:#x} expected {correct:#x}")

matrix_hash()
